### PR TITLE
update preprocess.py to use vllm render endpoint

### PR DESCRIPTION
**Pull request by shanjiaz**
## Purpose
This was originally [PR 794](https://github.com/vllm-project/speculators/pull/794). Changes are previously reviewed and approved. This PR updated e2e tests to use updated data flow & format. 

## Tests
Ran smoke tests locally, all passed, all e2e test [run](https://github.com/neuralmagic/llm-compressor-testing/actions/runs/31413006254/job/93535310680) in progress.

- Text-only smoke/regression tests now use pre-tokenized data, bypassing the render endpoint entirely
- Multimodal tests and MTP finetuning test prepare data inside the vLLM server context using the render endpoint
- Regression tests like test_finetuning_sanity and test_training_only switched from legacy data format to newly generated hidden states. Legacy data can be fully removed
- `run_prepare_data` in `tests/e2e/utils.py` now has an optional `render_endpoint` parameter
- Deleted `test_regex_patterns.py` (regex masking removed) 

llm-compressor-testing can't pick up Ranran's branch. Ran tests locally:
nightly (13 passed 16 skipped):
```
======================================================================================================================== warnings summary ========================================================================================================================
.venv/lib64/python3.12/site-packages/torch/jit/_script.py:365: 14 warnings
  /home/shanjiaz/speculators/.venv/lib64/python3.12/site-packages/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

tests/e2e/smoke/test_finetuning_sanity.py: 55 warnings
tests/e2e/smoke/test_mtp_conversion_roundtrip.py: 2 warnings
tests/e2e/smoke/test_mtp_finetuning.py: 1 warning
tests/e2e/smoke/test_offline_training.py: 1 warning
  /home/shanjiaz/speculators/.venv/lib64/python3.12/site-packages/huggingface_hub/file_download.py:1855: DeprecationWarning: hf_xet.download_files() is deprecated. Use XetSession().new_file_download_group().start_download_file() instead.
    xet_get(

tests/e2e/smoke/test_finetuning_sanity.py::test_finetuning_weight_sanity
  /home/shanjiaz/speculators/tests/e2e/smoke/test_finetuning_sanity.py:53: UserWarning: --target-layer-ids is not explicitly set. Setting target layers to [2, 16, 29]. If custom target layers were used when launching vllm datagen, please set them explicitly.
    model = Eagle3DraftModel.from_pretrained(PRETRAINED)

tests/e2e/smoke/test_mtp_conversion_roundtrip.py::test_mtp_roundtrip
  /home/shanjiaz/speculators/.venv/lib64/python3.12/site-packages/torch/_inductor/compile_fx.py:322: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
    warnings.warn(

tests/e2e/smoke/test_mtp_conversion_roundtrip.py::test_mtp_roundtrip
  /home/shanjiaz/speculators/.venv/lib64/python3.12/site-packages/torch/_inductor/lowering.py:7836: UserWarning: 
  Online softmax is disabled on the fly since Inductor decides to
  split the reduction. Cut an issue to PyTorch if this is an
  important use case and you want to speed it up with online
  softmax.
  
    warnings.warn(

tests/e2e/smoke/test_render_boundary.py::test_render_boundary_masks_against_live_vllm
tests/e2e/smoke/test_render_boundary.py::test_render_boundary_masks_against_live_vllm
  /home/shanjiaz/speculators/.venv/lib64/python3.12/site-packages/multiprocess/popen_fork.py:66: DeprecationWarning: This process (pid=1192357) is multi-threaded, use of fork() may lead to deadlocks in the child.
    self.pid = os.fork()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==================================================================================================== 13 passed, 16 skipped, 78 warnings in 2167.45s (0:36:07) ====================================================================================================
```

## Checklist

I have filled in:

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan/results, such as providing test command and pasting the results.
- [x] (Optional) The necessary documentation update.
- [x] I (a human) have written or reviewed the code in this pr to the best of my ability.

### CODE REVIEW

File: src/speculators/data_generation/render_client.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    body = {
+        "messages": messages,
+        "add_generation_prompt": add_generation_prompt,
+    }
+    if tools is not None:
+        body["tools"] = tools
+    if chat_template_kwargs is not None:
+        body["chat_template_kwargs"] = chat_template_kwargs
+
+    resp = httpx.post(url, json=body, timeout=timeout)
```
[the remark this answers is not in the document]
Comment (orestis-z) [reply src/speculators/data_generation/render_client.py:53]: this pr only modifies tests/* files, other stuff is from previously approved PR

File: src/speculators/data_generation/preprocessing.py [hunk 2]
```diff
@@ -500,56 +305,165 @@ def _parse_conv_tools(conv_tools: object, idx: int) -> list | None:
...
 ) -> dict[str, list]:
-    """Carry pre-tokenized ``(input_ids, loss_mask)`` rows through, truncated only.
+    """Carry speculator-format ``(input_ids, loss_mask)`` rows through.
 
-    On-policy regeneration already applied the boundary as the mask, so these rows
-    need no chat-template rendering or regex span detection.
+    The producer already recorded which target-model tokens are supervised, so
+    these rows only need truncation and filtering.
     """
     results: dict[str, list] = {"input_ids": [], "loss_mask": [], "seq_len": []}
```
[the remark this answers is not in the document]
Comment (orestis-z) [reply src/speculators/data_generation/preprocessing.py:421]: this pr only modifies tests/* files, other stuff is from previously approved PR

File: src/speculators/data_generation/preprocessing.py [hunk 3]
```diff
@@ -500,56 +305,165 @@ def _parse_conv_tools(conv_tools: object, idx: int) -> list | None:
...
+            render_endpoint,
+            max_length,
+            tools=parsed_tools,
+        )
+    # One row the render endpoint or boundary derivation can't handle must
+    # not kill the run. The failure modes can't be enumerated -- templates
+    # are swappable and raise arbitrary types -- so catch broadly and skip.
+    except Exception as e:
+        log.error(f"Failed to process conversation {idx}: {type(e).__name__}: {e}")
+        return []
```
[the remark this answers is not in the document]
Comment (orestis-z) [reply src/speculators/data_generation/preprocessing.py:336]: this pr only modifies tests/* files, other stuff is from previously approved PR

File: scripts/prepare_data.py [hunk 4]
```diff
@@ -125,12 +136,19 @@ def parse_args():
...
-            "If not provided, auto-detected from chat template."
+            "Base URL of a running vLLM server (e.g. http://localhost:8000). "
+            "The instance launched for hidden-state extraction serves this "
+            "too, so no second server is needed. Pass the base URL only: "
+            "/v1/chat/completions/render is appended to it, so the "
+            "/v1-suffixed form that data_generation_offline.py --endpoint "
+            "takes will 404. Conversations are tokenized by that endpoint and "
+            "the loss mask is derived from the render boundary. Rendering does "
+            "not generate responses or make arbitrary data on-policy. Required "
+            "unless every --data input already contains input_ids and loss_mask."
```
[the remark this answers is not in the document]
Comment (orestis-z) [reply scripts/prepare_data.py:151]: this pr only modifies tests/* files, other stuff is from previously approved PR

File: docs/user_guide/tutorials/train.md [hunk 5]
```diff
@@ -98,41 +98,39 @@ Note: if you are using an experiment tracker (e.g. trackio, wandb, tensorboard,
...
 ```
 
+The render endpoint applies the serving chat template, tokenizes each turn, and derives its loss mask. It does not generate responses or make a dataset on-policy, so the assistant responses must already come from the same target model and generation configuration used for training.
+
 **Parameters explained:**
 
 - `--model` - The target model you want to accelerate
-- `--data` - Dataset to use (built-in support for `sharegpt`, `ultrachat`. Otherwise provide a custom path to a jsonl file). Can be supplied multiple times to combine multiple datasets.
+- `--data` - On-policy target-model data, either natural-language `conversations` or speculator-format `input_ids` and `loss_mask`. Can be supplied multiple times to combine datasets.
+- `--render-endpoint` - Target model's vLLM base URL; required only for natural-language conversations.
```
[the remark this answers is not in the document]
Comment (orestis-z) [reply docs/user_guide/tutorials/train.md:133]: this pr only modifies tests/* files, other stuff is from previously approved PR

File: docs/cli/response_regeneration.md [hunk 6]
```diff
@@ -136,11 +136,11 @@ The text presets from the shared dataset registry (`DATASET_CONFIGS` in `specula
...
+The registry's multimodal preset, `sharegpt4v_coco`, is rejected because this regeneration pipeline cannot send its image content or retain it in a speculator-format row. Generate target responses with a multimodal-capable workflow, save the resulting natural-language conversations, and convert them with `prepare_data.py`.
 
 ## Output Format
 
-Rows are pre-tokenized and ready for training: one row per target generation, holding the prompt the target conditioned on followed by the tokens it generated. The endpoint must support `return_token_ids`, which the script uses to read the generation boundary directly instead of re-tokenizing the text and recovering the boundary with a regex.
+Rows are in speculator format and ready for training: one row per target generation, holding the prompt the target conditioned on followed by the tokens it generated. The endpoint must support `return_token_ids`, which the script uses to read the generation boundary directly instead of re-tokenizing the text and recovering the boundary with a regex.
```
[the remark this answers is not in the document]
Comment (orestis-z) [reply docs/cli/response_regeneration.md:143]: this pr only modifies tests/* files, other stuff is from previously approved PR

**Review by speculatorsbot — COMMENTED** [review]
## Review Summary

The non-test changes (source code, docs, CLI) carry over from the previously approved PR #794 and are not reviewed again here. This review focuses on the test-only changes authored in this PR.

### What the test changes do

1. **E2e tests switch from off-policy `"sharegpt"` to pre-tokenized on-policy datasets** (`hf:inference-optimization/speculators-ci-datasets:smoke_regen` / `tutorial_regen`). This aligns the test data with the new pipeline contract that requires target-model responses.

2. **Data preparation moves inside the vLLM server context** in `test_offline_training.py`, `test_online_training.py`, and `test_mtp_finetuning.py`. This is necessary because conversation-format datasets now need the render endpoint, which is served by the same vLLM instance used for hidden-state extraction. Pre-tokenized datasets pass through without rendering, so the endpoint is harmless when supplied but unnecessary.

3. **`test_resume_optimizer.py`** correctly omits `render_endpoint` since `smoke_regen` is pre-tokenized and no server is running at that point in the test.

4. **`test_training_only.py` and `test_finetuning_sanity.py`** consolidate onto `inference-optimization/speculators-ci-datasets`, replace `--legacy-data` with `--on-missing raise`, and add `--hidden-states-path` to match the current training script contract. The removal of `_generate_t2d_d2t` and the vocab mapping imports is consistent with the new data format bundling these.

5. **New `test_render_boundary.py` (e2e smoke)** exercises the live vLLM `/render` endpoint with Qwen3-0.6B. Good choice of model -- its template rewrites history (strips `<think>` from past turns), which forces the fan-out code path. The conversation is well-designed: a leading assistant turn (no row), two bounded assistant turns (one row each), and a trailing user turn (dropped). The assertions check the core invariant (user tokens never supervised, assistant tokens are) and verify per-row isolation and context re-rendering.

6. **New `test_render_boundary.py` (integration)** covers the edge cases a live server cannot produce on demand: scaffold LCP fallback (DeepSeek-R1 / Qwen3.5), the unstable-boundary guard, over-length turn skipping without dropping later turns, and the render client's error/retry paths. The `_patch_encode` helper is clean and the keying by `(prefix_len, add_generation_prompt)` makes the stubs easy to trace.

7. **Deleted tests** (`test_regex_patterns.py`, regex/HF-mask tests in `test_preprocessing.py`) are correct removals -- these tested `_detect_assistant_pattern`, `_create_loss_mask_from_offsets`, `_supports_assistant_mask`, and `_preprocess_batch` with the old signature, all of which are removed in the source changes.

8. **`test_response_regeneration.py`** updates `_preprocess_batch` call sites to use the new signature (`is_multimodal=False, render_endpoint=None` instead of `processor=None, assistant_pattern=None`). The passthrough contract is preserved.

9. **`tests/e2e/utils.py`** adds the optional `render_endpoint` parameter to `run_prepare_data`, forwarded as `--render-endpoint` when set. Clean and minimal.

### Observations

- The `hf:` spec `hf:inference-optimization/speculators-ci-datasets:smoke_regen` is parsed by `_load_hf_dataset` as `split="smoke_regen"` (no subset). CI confirms this resolves correctly. If the HF dataset is later reorganized to use named configs instead of split-level partitioning, these specs would need to become three-part (`hf:id:subset:train`). Not a blocker -- just noting the coupling.

- The deleted `test_load_and_preprocess_dataset_shuffles_combined_datasets` test verified that combined multi-dataset inputs are shuffled before `max_samples` truncation. That shuffle logic is unchanged in the new code, so coverage is only reduced, not broken. Worth noting if future refactors touch the combine-and-truncate path.

LGTM. The test updates are consistent, well-commented, and properly exercise both the new render-boundary pipeline and the speculator-format passthrough path.

**Comment by orestis-z** [comment]
e2e trigger (on identical branch on vllm-project): https://github.com/neuralmagic/llm-compressor-testing/actions/runs/31815657068

**Review by orestis-z — APPROVED** [review]
LGTM! 

once https://github.com/neuralmagic/llm-compressor-testing/actions/runs/31815657068 passes feel free to land.

**Comment by shanjiaz** [comment]
> LGTM!
> 
> once https://github.com/neuralmagic/llm-compressor-testing/actions/runs/31815657068 passes feel free to land.

Thanks for kicking this off! Merging since this all passed 🫡
